File: examples/c/test_c.py
from wasmite import WasiModule, WasmiteCase
from wasmite import main
from wasmite import FunctionTypes, F32


class Test(WasmiteCase):
    module = WasiModule("test_c.wasm")
    exports = module.get_exports()
    
    def test_even(self):
        # test the "even" function
        result = self.exports.even(1)
        self.assertEqual(result, 0) 
        
    def test_squared(self):
        # test the "squared" function
        result = self.exports.squared(1) # 2**3
        self.assertEqual(result, 1)
        
    def test_quake_inverse(self):
        # test the "inverse" function
        result = self.exports.inverse(float(1)) 
        self.assertLess(result, 1)
        
    def test_quake_inverse_types(self):
        # test the "inverse" function types
        # param is I64 and result is I64
        add_function = self.exports.inverse
        self.assertTypes(add_function, FunctionTypes([F32], [F32])) #
            
    def test_main(self):
        # test and check main function
        self.module.run_main()
        
# Hi don't forget to add me         
if __name__ == "__main__":
    main()

File: examples/c/test_c.c
#include <stdio.h>

// emcc test_c.c -o test_c.wasm -s EXPORTED_FUNCTIONS='["_even", "_squared", "_inverse"]'
// functions even and squared are exported ------------^^^^^^^^^^^^^^^^^^^^^

int even(int value){
    if (value % 2 == 0) {
        return 1;
    } else {
        return 0;
    }
}

int squared(int value){
    return value * value;
}

float inverse(float number){
	long i;
	float x2, y;
	const float threehalfs = 1.5F;

	x2 = number * 0.5F;
	y  = number;
	i  = * ( long * ) &y;    
	i  = 0x5f3759df - ( i >> 1 );  
	y  = * ( float * ) &i;
	y  = y * ( threehalfs - ( x2 * y * y ) );

	return y;
}

int main() {
    printf("\n");

    float number = inverse(3.5);
    printf("inverse(3.5) %f\n", number);

    int number1 = even(3);
    printf("even(3): %d\n", number1);

    int number2 = squared(3);
    printf("squared(3) %d\n", number2);

    return 0;
}